### PR TITLE
Introducing post authentication handler

### PR DESCRIPTION
**Pull request by nuwandiw**

### CODE REVIEW

**Comment by pulasthi7** [comment]
@nuwandiw : Squash and send to v5.6.x branch

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 1]
```diff
@@ -467,4 +472,14 @@ protected void handleAuthorization(HttpServletRequest request, HttpServletRespon
         }
     }
 
+    protected boolean isPostAuthenticationExtensionCompleted(AuthenticationContext context) {
+
+        Object object = context.getProperty(FrameworkConstants.POST_AUTHENTICATION_EXTENSION_COMPLETED);
+        if (object != null) {
+            return (Boolean) object;
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: Better to check for instanceof Boolean here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                       AuthenticationContext context) throws FrameworkException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Post authentication handling for missing claims started");
+        }
+
+        Object object = context.getProperty(FrameworkConstants.POST_AUTHENTICATION_REDIRECTION_TRIGGERED);
+        boolean postAuthRequestTriggered = false;
+        if (object != null) {
+            postAuthRequestTriggered = (boolean) object;
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java]: Better to check for instanceof Boolean here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                log.debug("Mandatory claims missing for the application : " + missingClaims);
+            }
+            //need to request for the missing claims before completing authentication
+            request.setAttribute(FrameworkConstants.RequestParams.FLOW_STATUS, AuthenticatorFlowStatus.INCOMPLETE);
+            context.setProperty(FrameworkConstants.POST_AUTHENTICATION_EXTENSION_COMPLETED, false);
+
+            try {
+                String queryString = FrameworkConstants.MISSING_CLAIMS + "=" + URLEncoder.encode(missingClaims);
+                queryString += "&" + FrameworkConstants.SESSION_DATA_KEY + "=" + context.getContextIdentifier();
+                queryString += "&" + "spName" + "=" +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java]: Can use the URIBuilder class

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        context.getSequenceConfig().getAuthenticatedUser().setUserAttributes(authenticatedUserAttributes);
+        context.setProperty(FrameworkConstants.POST_AUTHENTICATION_EXTENSION_COMPLETED, true);
+
+    }
+
+    private String getMissingClaims(Map<String,String> mappedAttrs, Map<String,String> requestedClaims) {
+
+        String missingClaims = "";
+        for (Map.Entry<String, String> entry : requestedClaims.entrySet())
+        {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java]: Move '{' to end of previous line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+            if (mappedAttrs.get(entry.getKey()) == null){
+                if (StringUtils.isNotBlank(missingClaims)) {
+                    missingClaims += ",";
+                }
+                missingClaims += entry.getKey();
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java]: Use StringBuilder

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java [hunk 6]
```diff
@@ -203,422 +186,333 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
...
+                sequenceConfig.setAuthenticatedIdPs(IdentityApplicationManagementUtil.getSignedJWT(
+                        jsonBuilder.toString(), sequenceConfig.getApplicationConfig()
+                                .getServiceProvider()));
 
-                Map<String, String> localIdpClaims = new HashMap<>();
-                for (Map.Entry<String, String> entry : claims.entrySet()) {
-                    String localClaim = claimMapping.get(entry.getKey());
-                    localIdpClaims.put(localClaim, entry.getValue());
+                if (!subjectFoundInStep) {
+                    stepConfig.setSubjectIdentifierStep(true);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java]: could be simplified to stepConfig.setSubjectIdentifierStep(!subjectFoundInStep)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java [hunk 7]
```diff
@@ -203,422 +186,333 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
...
-                        realm.getUserStoreManager().getSecondaryUserStoreManager(user.getUserStoreDomain());
-                userStoreManager.setUserClaimValues(user.getUserName(), localIdpClaims, null);
-            } catch (UserStoreException e) {
-                throw new FrameworkException(
-                        "Error while updating claims for local user. Could not update profile", e);
-            } catch (IdentityApplicationManagementException e) {
-                throw new FrameworkException(
-                        "Error while retrieving application claim mapping. Could not update profile", e);
+                if (!subjectAttributesFoundInStep) {
+                    stepConfig.setSubjectAttributeStep(true);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java]: could be simplified as above

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Mandatory claims missing for the application : " + missingClaims);
+            }
+            //need to request for the missing claims before completing authentication
+            request.setAttribute(FrameworkConstants.RequestParams.FLOW_STATUS, AuthenticatorFlowStatus.INCOMPLETE);
+            context.setProperty(FrameworkConstants.POST_AUTHENTICATION_EXTENSION_COMPLETED, false);
+
+            try {
+                String queryString = FrameworkConstants.MISSING_CLAIMS + "=" + URLEncoder.encode(missingClaims);
+                queryString += "&" + FrameworkConstants.SESSION_DATA_KEY + "=" + context.getContextIdentifier();
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultPostAuthenticationHandler.java]: shouldn't we encode this?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java [hunk 9]
```diff
@@ -203,422 +186,333 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
...
+                        if (context.getSequenceConfig().getApplicationConfig().getRequestedClaimMappings() == null ||
+                                context.getSequenceConfig().getApplicationConfig().getRequestedClaimMappings()
+                                        .isEmpty()) {
 
-                                if (localClaimValues != null && !localClaimValues.isEmpty()) {
-                                    mappedAttrs = localClaimValues;
-                                } else if (idpClaimValues != null && !idpClaimValues.isEmpty()) {
-                                    mappedAttrs = idpClaimValues;
-                                }
+                            if (localClaimValues != null && !localClaimValues.isEmpty()) {
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java]: Shall we use CollectionUtils.isNotEmpty() here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java [hunk 10]
```diff
@@ -203,422 +186,333 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
...
-                        if (roleAttr != null && roleAttr.trim().length() > 0) {
+                    String roleAttr = mappedAttrs.get(spRoleUri);
 
-                            String[] roles = roleAttr.split(",");
-                            mappedAttrs.put(
-                                    spRoleUri,
-                                    getServiceProviderMappedUserRoles(sequenceConfig,
-                                            Arrays.asList(roles)));
-                        }
+                    if (roleAttr != null && roleAttr.trim().length() > 0) {
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java]: StringUtils.isNotBlank() will be handy here.

**Comment by nuwandiw** [comment]
Closing PR since the fix is merged to v5.6.x branch with https://github.com/wso2/carbon-identity-framework/pull/652

**Pull request closed without merging**